REST-API-ADVANCED/advanced/src/main/java/com/epam/esm/security/TokenGenerator.java

```diff
@@ -0,0 +1,88 @@
+import org.springframework.security.core.Authentication;
+import org.springframework.security.oauth2.jwt.Jwt;
+import org.springframework.security.oauth2.jwt.JwtClaimsSet;
+import org.springframework.security.oauth2.jwt.JwtEncoder;
+import org.springframework.security.oauth2.jwt.JwtEncoderParameters;
+import org.springframework.stereotype.Component;
+
+import java.text.MessageFormat;
+import java.time.Duration;
+import java.time.Instant;
+import java.time.temporal.ChronoUnit;
+
+@Component
+public class TokenGenerator {
+
+    private final JwtEncoder accessTokenEncoder;
+
+    private final JwtEncoder refreshTokenEncoder;
+
+    @Autowired
+    public TokenGenerator(JwtEncoder accessTokenEncoder, JwtEncoder refreshTokenEncoder) {
+        this.accessTokenEncoder = accessTokenEncoder;
+        this.refreshTokenEncoder = refreshTokenEncoder;
+    }
+
+    private String createAccessToken(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        Instant now = Instant.now();
+        JwtClaimsSet jwtClaimsSet = JwtClaimsSet.builder()
+                .issuer("myApp")
+                .issuedAt(now)
+                .expiresAt(now.plus(5, ChronoUnit.MINUTES))
+                .subject(String.valueOf(user.getId()))
+                .claim("role", user.getRole())
+                .claim("provider", user.getProvider())
+                .build();
+        return accessTokenEncoder.encode(JwtEncoderParameters.from(jwtClaimsSet)).getTokenValue();
+    }
+
+    private String createRefreshToken(Authentication authentication) {
+        User user = (User) authentication.getPrincipal();
+        Instant now = Instant.now();
+        JwtClaimsSet jwtClaimsSet = JwtClaimsSet.builder()
+                .issuer("myApp")
+                .issuedAt(now)
+                .expiresAt(now.plus(30, ChronoUnit.DAYS))
+                .subject(String.valueOf(user.getId()))
+                .claim("role", user.getRole())
+                .claim("provider", user.getProvider())
+                .build();
+        return refreshTokenEncoder.encode(JwtEncoderParameters.from(jwtClaimsSet)).getTokenValue();
+    }
+
+    public TokenDTO createToken(Authentication authentication) {
+        if (!((authentication).getPrincipal() instanceof User user)) {
+            throw new BadCredentialsException(
+                    MessageFormat.format("principal {0} is not of User entity", authentication.getPrincipal().getClass())
+            );
+        }
+        TokenDTO tokenDTO = new TokenDTO();
+        tokenDTO.setUserId(String.valueOf(user.getId()));
+        tokenDTO.setAccessToken(createAccessToken(authentication));
+
+        String refreshToken;
+        if (authentication.getCredentials() instanceof Jwt jwt) {
+            Instant now = Instant.now();
+            Instant expiredAt = jwt.getExpiresAt();
+            Duration duration = Duration.between(now, expiredAt);
```

Comment:
IMO is readable enough
```suggestion
            Duration duration = Duration.between(Instant.now(), jwt.getExpiresAt());
```

Reply:
method removed:(